src/Storages/S3Queue/S3QueueOrderedFileMetadata.cpp

```diff
@@ -0,0 +1,257 @@
+#include "S3QueueOrderedFileMetadata.h"
+#include <Common/SipHash.h>
+#include <Common/getRandomASCIIString.h>
+#include <Common/logger_useful.h>
+#include <Common/DNSResolver.h>
+#include <Interpreters/Context.h>
+#include <Poco/JSON/JSON.h>
+#include <Poco/JSON/Object.h>
+#include <Poco/JSON/Parser.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    OrderedFileMetadata::Bucket getBucketForPathImpl(const std::string & path, size_t buckets_num)
+    {
+        return sipHash64(path) % buckets_num;
+    }
+
+    std::string getProcessedPath(const std::filesystem::path & zk_path, const std::string & path, size_t buckets_num)
+    {
+        if (buckets_num > 1)
+            return zk_path / "buckets" / toString(getBucketForPathImpl(path, buckets_num)) / "processed";
+        else
+            return zk_path / "processed";
+    }
+
+    zkutil::ZooKeeperPtr getZooKeeper()
+    {
+        return Context::getGlobalContextInstance()->getZooKeeper();
+    }
+}
+
+struct OrderedFileMetadata::BucketHolder
+{
+    BucketHolder(const std::string & bucket_lock_path_, zkutil::ZooKeeperPtr zk_client_)
+        : bucket_lock_path(bucket_lock_path_), zk_client(zk_client_) {}
+
+    void release()
+    {
+        if (released)
+            return;
+        released = true;
```

Comment:
If we fail - it is either keeper session expired or some bug. In the first case of expired session it does not make sense to retry (and it is normal behaviour). In this particular case we cannot distinguish a failure because of a normal case of keeper session expired and if we failed because of some bug, because keeper session could fail and restore again before we got into this method. So we cannot add any assertions here. And so it does not make sense to not set `released = true` and let it be retried in destructor. The error from `release` method is anyway rethrown - we call `release` explicitly in the code of Source.